fails to build due to safe-string warnings

I got this while trying to go from `opam.1.2.0` to opam2, by install `opam-devel`.

I believe it's because I'm using `4.06.1+trunk+force-safe-string` since that makes the `String_val` macro expect a `const char*`, not `char*`, which it's not getting from ml_solver ... but I'm not 100% that's the correct diagnosis.

```
#=== ERROR while installing mccs.1.1+8 ========================================#
# opam-version 1.2.2
# os           linux
# command      jbuilder build -p mccs
# path         /home/user/.opam/4.06.1+trunk+force-safe-string/build/mccs.1.1+8
# compiler     4.06.1+trunk+force-safe-string
# exit-code    1
# env-file     /home/user/.opam/4.06.1+trunk+force-safe-string/build/mccs.1.1+8/mccs-24530-ed04f4.env
# stdout-file  /home/user/.opam/4.06.1+trunk+force-safe-string/build/mccs.1.1+8/mccs-24530-ed04f4.out
# stderr-file  /home/user/.opam/4.06.1+trunk+force-safe-string/build/mccs.1.1+8/mccs-24530-ed04f4.err
### stderr ###
# [...]
# mccs_stubs.cpp:397:15: note: in expansion of macro ‘String_val’
#          dft = String_val(Some_val(Field(arg, 1)));
#                ^~~~~~~~~~
# mccs_stubs.cpp: In function ‘Solver ml2c_solver(value)’:
# /home/user/.opam/4.06.1+trunk+force-safe-string/lib/ocaml/caml/mlvalues.h:256:24: error: invalid conversion from ‘const char*’ to ‘char*’ [-fpermissive]
#  #define String_val(x) ((const char *) Bp_val(x))
#                        ~^~~~~~~~~~~~~~~~~~~~~~~~~
# mccs_stubs.cpp:438:23: note: in expansion of macro ‘String_val’
#        ret.lp_solver = String_val (Field (ml_solver, 1));
#                        ^~~~~~~~~~



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install opam-client 2.0.0
  ∗  install opam-devel  2.0.0
  ∗  install opam-solver 2.0.0
The following actions failed
  ∗  install mccs 1.1+8
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -34,6 +34,8 @@ matrix:
     env: OCAML_VERSION=4.07 MCCS_BACKENDS="GLPK CBC CLP SYMPHONY" MCCS_LPSOLVER=/usr/share/mccs/cbclp
   - os: linux
     env: OCAML_VERSION=4.08 MCCS_BACKENDS="GLPK CBC CLP SYMPHONY" MCCS_LPSOLVER=/usr/share/mccs/cbclp
+  - os: linux
+    env: OCAML_VERSION=4.08 OPAM_SWITCH=ocaml-variants.4.08.0+force-safe-string MCCS_BACKENDS="GLPK CBC CLP SYMPHONY" MCCS_LPSOLVER=/usr/share/mccs/cbclp
 addons:
   apt:
     update: true

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,5 +1,6 @@
 sudo: required
 language: c
+dist: bionic
 install: wget https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.travis-opam.sh
 before_install:
   - $TRAVIS_BUILD_DIR/.travis-ci.sh

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,6 +5,10 @@ install: wget https://raw.githubusercontent.com/ocaml/ocaml-ci-scripts/master/.t
 before_install:
   - $TRAVIS_BUILD_DIR/.travis-ci.sh
 script: bash -ex .travis-opam.sh
+env:
+  global:
+    - FORK_USER=dra27
+    - FORK_BRANCH=fix-bubblewrap
 matrix:
   include:
   - os: osx

--- a/src/abstract_solver.h
+++ b/src/abstract_solver.h
@@ -84,7 +84,7 @@ class abstract_solver {
 
   // ******************************************************************
   // write the internal representation of the problem to a file
-  // virtual int writelp(char *filename) { return 0; };
+  // virtual int writelp(const char *filename) { return 0; };
 
   // ******************************************************************
   // solve the problem (must return a value > 0 if a solution has been found, 0

--- a/src/count_criteria.h
+++ b/src/count_criteria.h
@@ -22,7 +22,7 @@ class count_criteria: public abstract_criteria {
   CUDFproblem *problem;      // a pointer to the problem
   abstract_solver *solver;   // a pointer to the solver
 
-  char *property_name;       // name of the property
+  const char *property_name; // name of the property
   bool has_property;         // is the property available ?
   // list of all versioned packages which have the property
   CUDFVersionedPackageList *versioned_pkg_with_property;
@@ -61,13 +61,13 @@ class count_criteria: public abstract_criteria {
   Count_scope scope;
 
   // Criteria initialization
-  count_criteria(char *property_name) { 
+  count_criteria(const char *property_name) {
     this->property_name = property_name;
     this->lambda_crit = +1; 
   };
 
   // Criteria initialization
-  count_criteria(char *property_name, Count_scope scope, CUDFcoefficient lambda_crit) { 
+  count_criteria(const char *property_name, Count_scope scope, CUDFcoefficient lambda_crit) {
     this->property_name = property_name;
     this->lambda_crit = lambda_crit; 
     this->scope = scope;

--- a/src/cudf.h
+++ b/src/cudf.h
@@ -74,9 +74,9 @@ extern CUDFVpkg *vpkg_false;
 // A package i.e. either a versioned package or a virtual package
 class CUDFPackage {
  public:
-  char *name;              // package name as provided by the CUDF problem file
-  int rank;                // rank of the package, i.e., column number in a simplex matrix or variable id
-  char *versioned_name;    // internal name of the package
+  const char *name;           // package name as provided by the CUDF problem file
+  int rank;                   // rank of the package, i.e., column number in a simplex matrix or variable id
+  const char *versioned_name; // internal name of the package
 
   bool in_reduced;
 
@@ -188,7 +188,7 @@ class CUDFPropertyValue {
  public:
   CUDFProperty *property;           // type of the property value
   int intval;                       // use to store property value when its basic type is an int
-  char *strval;                     // use to store property value when its basic type is a string
+  const char *strval;               // use to store property value when its basic type is a string
   CUDFVpkg *vpkg;                   // use to store property value when its basic type is a vpkg
   CUDFVpkgList *vpkglist;           // use to store property value when its basic type is a vpkglist
   CUDFVpkgFormula *vpkgformula;     // use to store property value when its basic type is a vpkgformula
@@ -205,7 +205,7 @@ class CUDFPropertyValue {
 // Class to describe user defined properties
 class CUDFProperty {
  public:
-  char *name;                         // property name
+  const char *name;                   // property name
 
   CUDFPropertyType type_id;           // property type
   CUDFEnums *enuml;                   // allowed enum values for enum type properties
@@ -275,7 +275,7 @@ public:
 extern int parse_cudf(FILE *input_file);
 
 /* // parse the CUDF problem from a string */
-extern int parse_cudf_string(char *str);
+extern int parse_cudf_string(const char *str);
 
 // operations to compare a package version to another one
 extern bool op_none_comp(CUDFVersion v1, CUDFVersion v2);

--- a/src/cudf.h
+++ b/src/cudf.h
@@ -163,11 +163,11 @@ typedef vector<CUDFVirtualPackage *> CUDFVirtualPackageList;
 typedef CUDFVirtualPackageList::iterator CUDFVirtualPackageListIterator;
 
 // type of an enums and its iterator type
-typedef vector<char *> CUDFEnums;
-typedef vector<char *>::iterator CUDFEnumsIterator;
+typedef vector<const char *> CUDFEnums;
+typedef vector<const char *>::iterator CUDFEnumsIterator;
 
 // get enum "estr" from "e" enum list
-extern char *get_enum(CUDFEnums *e, char *estr);
+extern const char *get_enum(CUDFEnums *e, const char *estr);
 
 
 // a property mapping type and its iterator type
@@ -194,7 +194,7 @@ class CUDFPropertyValue {
   CUDFVpkgFormula *vpkgformula;     // use to store property value when its basic type is a vpkgformula
 
   CUDFPropertyValue(CUDFProperty *the_property, int the_value);              // constructor for int valued properties
-  CUDFPropertyValue(CUDFProperty *the_property, char *the_value);            // constructor for string valued properties
+  CUDFPropertyValue(CUDFProperty *the_property, const char *the_value);      // constructor for string valued properties
   CUDFPropertyValue(CUDFProperty *the_property, CUDFVpkg *the_value);        // constructor for vpkg valued properties
   CUDFPropertyValue(CUDFProperty *the_property, CUDFVpkgList *the_value);    // constructor for vpkglist valued properties
   CUDFPropertyValue(CUDFProperty *the_property, CUDFVpkgFormula *the_value); // constructor for vpkgformula valued properties
@@ -214,14 +214,14 @@ class CUDFProperty {
 
   CUDFPropertyValue *default_value;   // default property value
 
-  CUDFProperty(char *tname, CUDFPropertyType ttype);                                 // CUDF property constructors (without and with default value)
-  CUDFProperty(char *tname, CUDFPropertyType ttype, int tdefault);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, char *tdefault);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum, char *tident);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefault);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkgList *tdefault);
-  CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkgFormula *tdefault);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype);                           // CUDF property constructors (without and with default value)
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, int tdefault);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, const char *tdefault);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum, const char *tident);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefault);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgList *tdefault);
+  CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgFormula *tdefault);
 
   ~CUDFProperty();
 

--- a/src/cudf_tools.cpp
+++ b/src/cudf_tools.cpp
@@ -117,7 +117,7 @@ CUDFVirtualPackage::~CUDFVirtualPackage() {
 
 // User property constructor
 // requires property name and type of the property
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -135,7 +135,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype) {
 
 // User property constructor
 // requires property name, type of the property (must be an int subtype) and int default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, int tdefault) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, int tdefault) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -169,8 +169,8 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, int tdefault) {
 CUDFProperty::~CUDFProperty() {
   free(name);
   if (type_id == pt_enum) {
-    for (vector<char*>::iterator it = enuml->begin(); it != enuml->end(); it++)
-      free(*it);
+    for (vector<const char*>::iterator it = enuml->begin(); it != enuml->end(); it++)
+      free((char*)*it);
     delete enuml;
   }
   delete default_value;
@@ -178,7 +178,7 @@ CUDFProperty::~CUDFProperty() {
 
 // User property constructor
 // requires property name, type of the property (must be a string subtype) and string default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, char *tdefault) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, const char *tdefault) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -194,7 +194,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, char *tdefault) 
 
 // User property constructor
 // requires property name, type of the property (must be a enum type) and enum default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -213,7 +213,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum
 
 // User property constructor
 // requires property name, type of the property (must be a enum type), the list of allowed enum values and enum default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum, char *tident) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum, const char *tident) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -227,7 +227,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum
 
   enuml = tenum;
 
-  char *defval = get_enum(tenum, tident);
+  const char *defval = get_enum(tenum, tident);
   if (defval == (char *)NULL) {
     PRINT_ERR("CUDF error: property %s default value can not be %s.\n", tname, tident);
     exit(-1);
@@ -238,7 +238,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFEnums *tenum
 
 // User property constructor
 // requires property name, type of the property (must be a vpkg type) and vpkg default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefault) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefault) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -255,7 +255,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefau
 
 // User property constructor
 // requires property name, type of the property (must be a vpkglist type) and vpkglist default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkgList *tdefault) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgList *tdefault) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -272,7 +272,7 @@ CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkgList *td
 
 // User property constructor
 // requires property name, type of the property (must be a vpkgformula type) and vpkgformula default value
-CUDFProperty::CUDFProperty(char *tname, CUDFPropertyType ttype, CUDFVpkgFormula *tdefault) {
+CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgFormula *tdefault) {
   int lgth = strlen(tname);
 
   if ((name = (char *)malloc(lgth+1)) == NULL) {
@@ -296,7 +296,7 @@ CUDFPropertyValue::CUDFPropertyValue(CUDFProperty *the_property, int the_value) 
 
 // User property value constructor
 // requires a pointer to the user property (must be a string subtype) and its string value
-CUDFPropertyValue::CUDFPropertyValue(CUDFProperty *the_property, char *the_value) {
+CUDFPropertyValue::CUDFPropertyValue(CUDFProperty *the_property, const char *the_value) {
   char *the_nvalue = (char *)malloc(strlen(the_value)+1);
 
   property = the_property;

--- a/src/cudf_tools.cpp
+++ b/src/cudf_tools.cpp
@@ -20,9 +20,9 @@ CUDFVpkg *vpkg_true = new CUDFVpkg((CUDFVirtualPackage *)NULL, op_none, 0);
 CUDFVpkg *vpkg_false = new CUDFVpkg((CUDFVirtualPackage *)NULL, op_none, 0);
 
 CUDFPackage::~CUDFPackage() {
-  free(name);
+  free((char *)name);
   if (versioned_name != NULL && versioned_name != name)
-    free(versioned_name);
+    free((char *)versioned_name);
 }
 
 // Versioned package constructor
@@ -30,11 +30,10 @@ CUDFPackage::~CUDFPackage() {
 CUDFVersionedPackage::CUDFVersionedPackage(const char *pkg_name, int my_rank) {
 
   // Get some piece of memory to store package name
-  if ((name = (char *)malloc(strlen(pkg_name)+1)) == NULL) {
+  if (!(name = strdup(pkg_name))) {
     PRINT_ERR("error: cannot alloc name for CUDFVersionedPackage.\n");
     exit(-1);
   }
-  strcpy(name, pkg_name);
   
   // Default initialization of variables
   versioned_name = (char *)NULL;
@@ -83,11 +82,11 @@ void CUDFVersionedPackage::set_version(CUDFVersion pkg_version) {
   static char temp[50];
 
   sprintf(temp, "%" CUDFint64"u", pkg_version);
-  if ((versioned_name = (char *)malloc(strlen(name)+strlen(temp)+2)) == NULL) {
+  if ((versioned_name = (const char *)malloc(strlen(name)+strlen(temp)+2)) == NULL) {
     PRINT_ERR("error: cannot alloc versioned_name for CUDFVersionedPackage.\n");
     exit(-1);
   }
-  sprintf(versioned_name, "%s_%s", name, temp);
+  sprintf((char *)versioned_name, "%s_%s", name, temp);
     
   version = pkg_version;
 }
@@ -95,11 +94,10 @@ void CUDFVersionedPackage::set_version(CUDFVersion pkg_version) {
 // Virtual package constructor
 // requires virtual package name and virtual package rank
 CUDFVirtualPackage::CUDFVirtualPackage(const char *pkg_name, int my_rank) {    
-  if ((name = (char *)malloc(strlen(pkg_name)+1)) == NULL) {
+  if (!(name = strdup(pkg_name))) {
     PRINT_ERR("error: cannot alloc name for CUDFVirtualPackage.\n");
     exit(-1);
   }
-  strcpy(name, pkg_name);
   versioned_name = name;
   
   highest_installed = (CUDFVersionedPackage *)NULL;
@@ -118,13 +116,10 @@ CUDFVirtualPackage::~CUDFVirtualPackage() {
 // User property constructor
 // requires property name and type of the property
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = true;
@@ -136,13 +131,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype) {
 // User property constructor
 // requires property name, type of the property (must be an int subtype) and int default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, int tdefault) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = false;
@@ -167,7 +159,7 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, int tdefau
 }
 
 CUDFProperty::~CUDFProperty() {
-  free(name);
+  free((char *)name);
   if (type_id == pt_enum) {
     for (vector<const char*>::iterator it = enuml->begin(); it != enuml->end(); it++)
       free((char*)*it);
@@ -179,13 +171,10 @@ CUDFProperty::~CUDFProperty() {
 // User property constructor
 // requires property name, type of the property (must be a string subtype) and string default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, const char *tdefault) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = false;
@@ -195,13 +184,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, const char
 // User property constructor
 // requires property name, type of the property (must be a enum type) and enum default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = true;
@@ -214,13 +200,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums 
 // User property constructor
 // requires property name, type of the property (must be a enum type), the list of allowed enum values and enum default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums *tenum, const char *tident) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = true;
@@ -228,7 +211,7 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums 
   enuml = tenum;
 
   const char *defval = get_enum(tenum, tident);
-  if (defval == (char *)NULL) {
+  if (defval == NULL) {
     PRINT_ERR("CUDF error: property %s default value can not be %s.\n", tname, tident);
     exit(-1);
   } else
@@ -239,13 +222,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFEnums 
 // User property constructor
 // requires property name, type of the property (must be a vpkg type) and vpkg default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkg *tdefault) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = false;
@@ -256,13 +236,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkg *
 // User property constructor
 // requires property name, type of the property (must be a vpkglist type) and vpkglist default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgList *tdefault) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = false;
@@ -273,13 +250,10 @@ CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgLi
 // User property constructor
 // requires property name, type of the property (must be a vpkgformula type) and vpkgformula default value
 CUDFProperty::CUDFProperty(const char *tname, CUDFPropertyType ttype, CUDFVpkgFormula *tdefault) {
-  int lgth = strlen(tname);
-
-  if ((name = (char *)malloc(lgth+1)) == NULL) {
+  if (!(name = strdup(tname))) {
     PRINT_ERR("error: cannot alloc name for property %s.\n", tname);
     exit(-1);
   }
-  strcpy(name, tname);
     
   type_id = ttype;
   required = false;
@@ -327,7 +301,7 @@ CUDFPropertyValue::CUDFPropertyValue(CUDFProperty *the_property, CUDFVpkgFormula
 
 CUDFPropertyValue::~CUDFPropertyValue() {
   switch (property->type_id) {
-  case pt_string: free(strval); break;
+  case pt_string: free((char *)strval); break;
   case pt_vpkg: case pt_veqpkg: delete vpkg; break;
   case pt_vpkglist: case pt_veqpkglist: delete vpkglist; break;
   case pt_vpkgformula: delete vpkgformula; break;

--- a/src/glpk_solver.cpp
+++ b/src/glpk_solver.cpp
@@ -59,7 +59,7 @@ int glpk_solver::set_intvar_range(int rank, CUDFcoefficient lower, CUDFcoefficie
 }
 
 // write the problem into a file
-// int glpk_solver::writelp(char *filename) { glp_write_lp(lp, NULL, filename); return 0; }
+// int glpk_solver::writelp(const char *filename) { glp_write_lp(lp, NULL, filename); return 0; }
 
 void glpk_solver::abort(void) {
   this->aborted = true;

--- a/src/glpk_solver.h
+++ b/src/glpk_solver.h
@@ -25,7 +25,7 @@ class glpk_solver: public abstract_solver, public scoeff_solver<double, 1, 1>  {
   int set_intvar_range(int rank, CUDFcoefficient lower, CUDFcoefficient upper);
 
   // Write the lp on a file
-  // int writelp(char *filename);
+  // int writelp(const char *filename);
 
   // Solve the problem
   int solve();

--- a/src/lp_solver.cpp
+++ b/src/lp_solver.cpp
@@ -85,7 +85,7 @@ int lp_solver::init_solver(CUDFVersionedPackageList *all_versioned_packages, int
 }
 
 // write the problem into a file
-int lp_solver::writelp(char *filename) { return 0; }
+int lp_solver::writelp(const char *filename) { return 0; }
 
 // solve the current problem
 int lp_solver::solve() {

--- a/src/lp_solver.cpp
+++ b/src/lp_solver.cpp
@@ -27,7 +27,7 @@ static unsigned long pid = 0;
 static unsigned long uid = 0;
 
 // external function for solver creation
-abstract_solver *new_lp_solver(char *lpsolver) { return new lp_solver(lpsolver); }
+abstract_solver *new_lp_solver(const char *lpsolver) { return new lp_solver(lpsolver); }
 
 // solver initialisation
 int lp_solver::init_solver(CUDFVersionedPackageList *all_versioned_packages, int other_vars) {

--- a/src/lp_solver.h
+++ b/src/lp_solver.h
@@ -92,12 +92,12 @@ class lp_solver: public abstract_solver, public scoeff_solver<CUDFcoefficient, 0
   char lpoutfilename[256];
   FILE *lpfile, *ctlpfile;
 
-  char *lpsolver; // name of the solver to call
+  const char *lpsolver; // name of the solver to call
 
   char mult;
 
   // solver creation
-  lp_solver(char *lpsolver) {
+  lp_solver(const char *lpsolver) {
     this->lpsolver = lpsolver;
     nb_packages = 0;
     all_versioned_packages = (CUDFVersionedPackageList *)NULL;

--- a/src/lp_solver.h
+++ b/src/lp_solver.h
@@ -18,7 +18,7 @@ class lp_solver: public abstract_solver, public scoeff_solver<CUDFcoefficient, 0
   // Solver initialization
   int init_solver(CUDFVersionedPackageList *all_versioned_packages, int other_vars);
   // Write the lp on a file
-  int writelp(char *filename);
+  int writelp(const char *filename);
 
   // Solve the problem
   int solve();

--- a/src/mccs_stubs.cpp
+++ b/src/mccs_stubs.cpp
@@ -148,7 +148,7 @@ CUDFPropertyType ml2c_propertytype(value pt)
 
 CUDFVpkg * ml2c_vpkg(Virtual_packages * tbl, value ml_vpkg)
 {
-  char * name = String_val(Field(ml_vpkg, 0));
+  const char * name = String_val(Field(ml_vpkg, 0));
   CUDFVirtualPackage * virt = tbl->get(name);
   value constr_opt = Field(ml_vpkg, 1);
   if (constr_opt == Val_none) return new CUDFVpkg(virt, op_none, 0);
@@ -219,7 +219,7 @@ value c2ml_vpkgformula(CUDFVpkgFormula * form)
 
 CUDFPropertyValue * ml2c_property(Virtual_packages * tbl, CUDFProperties * properties, value ml_prop)
 {
-  char * prop_name = String_val(Field(ml_prop,0));
+  const char * prop_name = String_val(Field(ml_prop,0));
   CUDFPropertiesIterator prop_it;
   CUDFProperty * prop;
   value ml_v = Field(ml_prop,1);
@@ -304,7 +304,7 @@ value c2ml_propertylist (CUDFPropertyValueList * plist)
 
 CUDFVersionedPackage * ml2c_package(Virtual_packages * tbl, CUDFProperties * properties, int &max_rank, value ml_package)
 {
-  char * package = String_val(Field(ml_package, 0));
+  const char * package = String_val(Field(ml_package, 0));
   unsigned int version = Int_val(Field(ml_package, 1));
   bool installed = Bool_val(Field(ml_package, 5));
 
@@ -373,7 +373,7 @@ value c2ml_package(CUDFVersionedPackage * pkg)
 
 CUDFProperty * ml2c_propertydef(Virtual_packages * tbl, value ml_pdef)
 {
-  char * property_name = String_val(Field(ml_pdef,0));
+  const char * property_name = String_val(Field(ml_pdef,0));
   value def = Field(ml_pdef,1);
   CUDFPropertyType ty = ml2c_propertytype(Field(def,0));
   value arg = Field(def,1);
@@ -390,7 +390,7 @@ CUDFProperty * ml2c_propertydef(Virtual_packages * tbl, value ml_pdef)
       p = new CUDFProperty(property_name, ty, String_val(Some_val(arg))); break;
     case pt_enum: {
       CUDFEnums * enuml = new CUDFEnums;
-      char * dft;
+      const char * dft;
       for (value l = Field(arg, 0); l != Val_emptylist; l = Field(l, 1))
         enuml->push_back(String_val(Field(l,0)));
       if (Field(arg, 1) == Val_none)
@@ -454,10 +454,10 @@ Solver ml2c_solver(value ml_solver)
 }
 
 // get an enum from its name in an enum list
-char *get_enum(CUDFEnums *e, char *estr) {
+const char *get_enum(CUDFEnums *e, const char *estr) {
   for (CUDFEnumsIterator ei = e->begin(); ei != e->end(); ei++)
     if (strcmp((*ei), estr) == 0) return (*ei);
-  return (char *)NULL;
+  return (const char *)NULL;
 }
 
 typedef struct {

--- a/src/mccscudf.cpp
+++ b/src/mccscudf.cpp
@@ -18,7 +18,7 @@
 
 // underlying solver declaration
 // allows using solvers withour having to include the whole solver classes
-extern abstract_solver *new_lp_solver(char *lpsolver);
+extern abstract_solver *new_lp_solver(const char *lpsolver);
 /* extern abstract_solver *new_pblib_solver(char *pbsolver); */
 //extern abstract_solver *new_ampl_solver(char *amplsolver);
 #ifdef USECPLEX 

--- a/src/mccscudf.h
+++ b/src/mccscudf.h
@@ -6,7 +6,7 @@ typedef enum {CPLEX, GUROBI, LPSOLVE, GLPK, LP, CLP, CBC, SYMPHONY} Solver_backe
 
 typedef struct {
   Solver_backend backend;
-  char * lp_solver;
+  const char * lp_solver;
 } Solver;
 
 typedef struct {

--- a/src/osi_solver.h
+++ b/src/osi_solver.h
@@ -36,7 +36,7 @@ class osi_solver: public abstract_solver  {
   int set_intvar_range(int rank, CUDFcoefficient lower, CUDFcoefficient upper);
 
   // Write the lp on a file
-  // int writelp(char *filename);
+  // int writelp(const char *filename);
 
   // Solve the problem
   int solve();
@@ -223,7 +223,7 @@ int osi_solver<OsiSolver>::end_objectives(void) {
 }
 
 // write the problem into a file
-// int osi_solver<OsiSolver>::writelp(char *filename) { glp_write_lp(lp, NULL, filename); return 0; }
+// int osi_solver<OsiSolver>::writelp(const char *filename) { glp_write_lp(lp, NULL, filename); return 0; }
 
 template<class OsiSolver>
 void osi_solver<OsiSolver>::abort(void) {

